### PR TITLE
[WIP] Fix compatibility with DSP 0.9

### PR DESCRIPTION
**Pull request by PhantomGamers**
We need to see what broke after the update. here is a quick list of things that come to my mind atm.

- Belts are at least partially broken. picking up items manually does not pick them up for other players, outbound belts from advanced miners do not output stuff for others until they _try_ to build the belts themself, ...
- ILS code has been changed (maybe just a bit?), but its hard to update our transpiler to match the new code. Maybe its easier to rewrite it.
- did not have a look yet at the logic but as the new item enhancement seems to give % chance boosts (afaik) maybe it could be challenging to sync that?
- generally sync the new facilities
- new mecha stuff

### CODE REVIEW

File: NebulaModel/DataStructures/PlayerData.cs [hunk 1]
```diff
@@ -45,6 +48,20 @@ public void Serialize(INetDataWriter writer)
...
+            writer.Put(Appearance != null);
+            if(Appearance != null)
+            {
+                using (MemoryStream ms = new MemoryStream())
+                {
+                    using (BinaryWriter wr = new BinaryWriter(ms))
+                    {
+                        Appearance.Export(wr);
+                    }
+                    byte[] export = ms.ToArray();
```
Comment (Danil0v3s) [inline NebulaModel/DataStructures/PlayerData.cs:60]: Isn't `Appearance.ToByte()` a shortcut to this?

File: NebulaModel/DataStructures/PlayerData.cs [hunk 2]
```diff
@@ -45,6 +48,20 @@ public void Serialize(INetDataWriter writer)
             Rotation.Serialize(writer);
             BodyRotation.Serialize(writer);
             Mecha.Serialize(writer);
+            writer.Put(Appearance != null);
```
Comment (Danil0v3s) [inline NebulaModel/DataStructures/PlayerData.cs:51]: I think you could avoid this line if you use the length to tell whether appearance is there or not

File: NebulaModel/DataStructures/PlayerData.cs [hunk 3]
```diff
@@ -63,6 +80,20 @@ public void Deserialize(INetDataReader reader)
...
+                int len = reader.GetInt();
+                byte[] data = new byte[len];
+                reader.GetBytes(data, len);
+                using (MemoryStream ms = new MemoryStream(data))
+                using (BinaryReader br = new BinaryReader(ms))
+                {
+                    Appearance = new MechaAppearance();
+                    Appearance.Init();
+                    Appearance.Import(br);
+                }
```
Comment (Danil0v3s) [inline NebulaModel/DataStructures/PlayerData.cs:95]: Can't you use `Appearance.FromBytes(data)`?

File: NebulaModel/DataStructures/PlayerData.cs (hunk 1 above)
Comment (sp00ktober) [reply NebulaModel/DataStructures/PlayerData.cs:60]: It is, could probably be exchanged but would have the same effect.

File: NebulaModel/DataStructures/PlayerData.cs (hunk 2 above)
Comment (sp00ktober) [reply NebulaModel/DataStructures/PlayerData.cs:51]: I just took the logic from here: https://github.com/hubastard/nebula/blob/3c7da1bd62a9e58d77eac2f898c07f2fc2d9ccec/NebulaModel/DataStructures/MechaData.cs#L48

File: NebulaModel/DataStructures/PlayerData.cs (hunk 3 above)
Comment (sp00ktober) [reply NebulaModel/DataStructures/PlayerData.cs:95]: Again, yes. But its just another way to write the same thing.

**Comment by sp00ktober** [comment]
Im putting a pre release version here and on our discord server now. Maybe we can catch some more bugs before the actual release :D

[nebula_0.8.0_pre-release.zip](https://github.com/hubastard/nebula/files/7997635/nebula_0.8.0_pre-release.zip)

**Comment by sp00ktober** [comment]
closes #230 
closes #229 
closes #210

**Comment by Jaymus3** [comment]
Tested the mod with a couple of people.  So far, everything seems to work properly in the starting system.  Got this error when one player disconnected on the other client that was still connected.  I've also attached the log from that client.  (If this isn't where you'd like pre-release issues reported, feel free to point me to where you'd rather.)

[LogOutput.log](https://github.com/hubastard/nebula/files/8020818/LogOutput.log)

**Comment by sp00ktober** [comment]
Totally fine to report here i guess.
There where some other people reporting this too, i wonder if we should handle undefined packets more gracefully

**Comment by HoshixD** [comment]
I did ~ 10h of testing with one of my friends. 
So far the mod works like a charm. Some feedback:

- We encountered the error of the post above several times (only host). Does not seem to break the game though.
- We encountered 3 instances of the game crashing the Laptop of my friend. We can't say whether that is due to the DSP-update or some changes you guys implemented. 
- There are still some ship artifacts left in orbit (only client) if a player is in transition between systems and the ships use warp cells. Definitely got better though...
- Duping seems to be fixed. We found no way to simulate the environment that would allow for the client to dupe items by manually placing them on a belt leading to a storage unit. This no longer works no matter the latency of the host/client. 

Thanks for the amount of work and time you put into this. Attached is the log file of the host. I can provide the client file if needed.
[LogOutput.log](https://github.com/hubastard/nebula/files/8024006/LogOutput.log)

**Review by sp00ktober — APPROVED** [review]

**Pull request merged**